src/backend/distributed/planner/multi_planner.c

```diff
@@ -294,176 +333,235 @@ CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query
 
 
 /*
- * GetMultiPlan returns the associated MultiPlan for a PlannedStmt if the
- * statement requires distributed execution, NULL otherwise.
+ * GetMultiPlan returns the associated MultiPlan for a CustomScan.
  */
 MultiPlan *
-GetMultiPlan(PlannedStmt *result)
+GetMultiPlan(CustomScan *customScan)
 {
-	char *serializedMultiPlan = NULL;
 	MultiPlan *multiPlan = NULL;
 
-	serializedMultiPlan = GetMultiPlanString(result);
-	multiPlan = (MultiPlan *) CitusStringToNode(serializedMultiPlan);
-	Assert(CitusIsA(multiPlan, MultiPlan));
-
-	return multiPlan;
-}
+	Assert(IsA(customScan, CustomScan));
+	Assert(list_length(customScan->custom_private) == 1);
 
+	multiPlan = DeSerializeMultiPlan(linitial(customScan->custom_private));
 
-/* Does the passed in statement require distributed execution? */
-bool
-HasCitusToplevelNode(PlannedStmt *result)
-{
-	/*
-	 * Can't be a distributed query if the extension hasn't been loaded
-	 * yet. Directly return false, part of the required infrastructure for
-	 * further checks might not be present.
-	 */
-	if (!CitusHasBeenLoaded())
-	{
-		return false;
-	}
-
-	if (GetMultiPlanString(result) == NULL)
-	{
-		return false;
-	}
-	else
-	{
-		return true;
-	}
+	return multiPlan;
 }
 
 
 /*
- * CreateCitusToplevelNode creates the top-level planTree node for a
- * distributed statement. That top-level node is a) recognizable by the
- * executor hooks, allowing them to redirect execution, b) contains the
- * parameters required for distributed execution.
- *
- * The exact representation of the top-level node is an implementation detail
- * which should not be referred to outside this file, as it's likely to become
- * version dependant. Use GetMultiPlan() and HasCitusToplevelNode() to access.
+ * SerializeMultiPlan returns the string representing the distributed plan in a
+ * Const node.
  *
- * Internally the data is stored as arguments to a 'citus_extradata_container'
- * function, which has to be removed from the really executed plan tree before
- * query execution.
+ * Note that this should be improved for 9.6+, we we can copy trees efficiently.
```

Comment:
I believe we loosely support last two major PostgreSQL versions, so we can make this change after PostgreSQL 10 is released.